### PR TITLE
Delete background states

### PR DESCRIPTION
**Pull request by typfel**
This intermediate PR delete the background states from the state machine, they are replaced by the `OperationStatus` which keep track of when the app is in `foreground`/`background` or doing a `call`, `background fetch` or `background task`.

Depends on:
- [x] https://github.com/wireapp/wire-ios-request-strategy/pull/23
- [x] https://github.com/wireapp/wire-ios-message-strategy/pull/51

### CODE REVIEW

File: Source/UserSession/SyncStatus.swift [hunk 1]
```diff
@@ -17,8 +17,11 @@
 //
 
 
+
+
+
```
Comment (vytis) [inline Source/UserSession/SyncStatus.swift]: Quite a bit of whitespace

File: Source/Synchronization/ApplicationStatusDirectory.swift [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+        apnsConfirmationStatus.tearDown()
+        clientRegistrationStatus.tearDown()
+        clientUpdateStatus.tearDown()
+    }
+    
+    public var deliveryConfirmation: DeliveryConfirmationDelegate {
+        get {
+            return apnsConfirmationStatus
```
Comment (vytis) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: Ca you simply return stuff here?

File: Source/Synchronization/ApplicationStatusDirectory.swift [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    public let userProfileUpdateStatus : UserProfileUpdateStatus
+    public let clientRegistrationStatus : ZMClientRegistrationStatus
+    public let clientUpdateStatus : ClientUpdateStatus
+    public let pingBackStatus : BackgroundAPNSPingBackStatus
+    public let accountStatus : ZMAccountStatus
+    public let proxiedRequestStatus : ProxiedRequestsStatus
+    public let syncStatus : SyncStatus
+    public let operationStatus : OperationStatus
+    public let requestCancellation: ZMRequestCancellation
+
```
Comment (vytis) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: Does it make sense to add the `UserProfileImageUpdateStatus` introduced for asset v3 in here also?

Comment (typfel) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: This branch is behind `develop` quite a bit. But yes I suppose I should add it here.

File: Source/Synchronization/ApplicationStatusDirectory.swift (hunk 2 above)
Comment (typfel) [inline Source/Synchronization/ApplicationStatusDirectory.swift]: let's try

File: Source/UserSession/OperationStatus.swift [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        backgroundFetchTimer?.invalidate()
+        backgroundFetchTimer = nil
+    }
+    
+    public func finishBackgroundTask(withTaskResult result: BackgroundTaskResult) {
+        DispatchQueue.main.async {
+            self.backgroundTaskHandler?(result)
+            self.backgroundTaskHandler = nil
+        }
+        backgroundTaskTimer?.invalidate()
```
Comment (marcoconti83) [inline Source/UserSession/OperationStatus.swift]: Should we invert the order and put the dispatch at the end, as with this code the handler might fire before we invalidate the timer? Not sure if that's a problem, but potentially we could call the handler and the timer could expire before we have a chance to call `invalidate()`

File: zmessaging-cocoa.xcodeproj/xcshareddata/xcschemes/zmessaging-ios.xcscheme [hunk 5]
```diff
@@ -113,8 +113,8 @@
          </EnvironmentVariable>
          <EnvironmentVariable
             key = "ZMLOG_TAGS"
-            value = "MockTransportRequests"
-            isEnabled = "NO">
+            value = "MockTransportRequests,Authentication,State machine"
+            isEnabled = "YES">
```
Comment (marcoconti83) [inline zmessaging-cocoa.xcodeproj/xcshareddata/xcschemes/zmessaging-ios.xcscheme]: This should be "NO"

File: zmessaging-cocoa.xcodeproj/xcshareddata/xcschemes/zmessaging-ios.xcscheme [hunk 6]
```diff
@@ -23,7 +23,7 @@
          <BuildActionEntry
             buildForTesting = "YES"
             buildForRunning = "NO"
-            buildForProfiling = "NO"
+            buildForProfiling = "YES"
```
Comment (marcoconti83) [inline zmessaging-cocoa.xcodeproj/xcshareddata/xcschemes/zmessaging-ios.xcscheme]: Is this needed?

Comment (typfel) [inline zmessaging-cocoa.xcodeproj/xcshareddata/xcschemes/zmessaging-ios.xcscheme]: No I will revert all xcscheme changes

File: Source/UserSession/OperationStatus.swift (hunk 4 above)
Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: Yep let's do that it sounds safer.

File: Source/UserSession/SyncStatus.swift [hunk 7]
```diff
@@ -124,28 +123,6 @@ extension SyncStatus {
 // MARK: Quick Sync
 extension SyncStatus {
     
-    public func didEnterForeground() {
```
Comment (basine) [inline Source/UserSession/SyncStatus.swift]: This is not needed anymore?

File: Source/UserSession/SyncStatus.swift [hunk 8]
```diff
@@ -16,9 +16,8 @@
 // along with this program. If not, see http://www.gnu.org/licenses/.
 //
 
-
 @objc public enum SyncPhase : Int {
-    case idle, fetchingLastUpdateEventID, fetchingConnections, fetchingConversations, fetchingUsers, fetchingMissedEvents, done
```
Comment (basine) [inline Source/UserSession/SyncStatus.swift]: So when you initialize the class, you will always be in fetchingLastUpdateEventID?

File: Source/UserSession/SyncStatus.swift (hunk 7 above)
Comment (typfel) [inline Source/UserSession/SyncStatus.swift]: The idea now is that synchronisation is not tied to being in the foreground since we might want to do it in the background while in a call. It will stop the synchronisation when going into the background since the websocket will be closed.

File: Source/UserSession/SyncStatus.swift (hunk 8 above)
Comment (typfel) [inline Source/UserSession/SyncStatus.swift]: can also be `fetchingMissingEvents` if has a previous lastUpdateEventID.

Comment (typfel) [inline Source/UserSession/SyncStatus.swift]: I thought I could remove the `idle`state since the strategies them self will decide if they are allowed to sync.

File: Source/UserSession/OperationStatus.swift [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        RequestAvailableNotification.notifyNewRequestsAvailable(self)
+    }
+    
+    public func startBackgroundTask(withCompletionHandler completionHandler: @escaping BackgroundTaskHandler) {
+        startBackgroundTask(timeout: 30.0, withCompletionHandler: completionHandler)
+    }
+    
+    public func startBackgroundTask(timeout: TimeInterval, withCompletionHandler completionHandler: @escaping BackgroundTaskHandler) {
+        guard backgroundTaskHandler == nil, isInBackground else {
+            return completionHandler(.failed)
```
Comment (basine) [inline Source/UserSession/OperationStatus.swift]: How does it work if you e.g. reply to two push notifications under bad network?

Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: Second will fail

File: Source/UserSession/OperationStatus.swift [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        backgroundTaskHandler = completionHandler
+        backgroundTaskTimer = Timer.scheduledTimer(timeInterval: timeout, target: self, selector: #selector(backgroundTaskTimeout), userInfo: nil, repeats: false)
+    }
+    
+    dynamic func backgroundFetchTimeout() {
+        finishBackgroundFetch(withFetchResult: .failed)
+    }
+    
+    dynamic func backgroundTaskTimeout() {
+        finishBackgroundTask(withTaskResult: .failed)
```
Comment (basine) [inline Source/UserSession/OperationStatus.swift]: I always wondered if we should also cancel all requests we are currently doing. This would be a different PR though.

File: Source/UserSession/SyncStatus.swift (hunk 8 above)
Comment (basine) [inline Source/UserSession/SyncStatus.swift]: Yeah they should "know" if they are authenticated

File: Source/UserSession/SyncStatus.swift (hunk 7 above)
Comment (basine) [inline Source/UserSession/SyncStatus.swift]: So you said, we stop the sync, don't we need to restart it at some point?

File: Source/UserSession/OperationStatus.swift (hunk 9 above)
Comment (basine) [inline Source/UserSession/OperationStatus.swift]: Would it be better to have a queue of tasks then?

File: Source/UserSession/OperationStatus.swift (hunk 10 above)
Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: Might be a good yes. Currently OperationStatus has no knowledge of request strategies which is nice.

File: Source/UserSession/OperationStatus.swift (hunk 9 above)
Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: Yep, might need to introduce an `task identifier` to separate the completion handlers if keep multiple around.

Comment (typfel) [inline Source/UserSession/OperationStatus.swift]: Queue doesn't really work since the request strategies wouldn't now which one is allowed to run.

File: Source/UserSession/SyncStatus.swift (hunk 7 above)
Comment (typfel) [inline Source/UserSession/SyncStatus.swift]: We still do when the websocket opens, no?

**Comment by typfel** [comment]
I'll merge this know and merge develop into all branches. Support for multiple background tasks will come in a separate PR.
